snuba/api.py

```diff
@@ -331,19 +333,22 @@ def parse_and_run_query(dataset, request: Request, timer):
     if max_days is not None and (to_date - from_date).days > max_days:
         from_date = to_date - timedelta(days=max_days)
 
-    request.query['conditions'].extend([
+    # TODO: replace all this logic with calls to the extensions query processors
+    # like:
+    # processors = dataset.get_query_processors()
```

Comment:
`s/get_query_processors/get_extensions`.

Separately, trying to make sure I understand the abstraction and had a few of questions wrt the call site:
1. Is the intention that we will get all the processors and loop over them? Or do something like:
    ```
    if 'project' in processors:
        query = processors[extension].process_query(query, request.extensions[extension])
    ```
1. It seems like some extensions rely on other extensions, like [here](https://github.com/getsentry/snuba/blob/master/snuba/api.py#L348-L350). There, we rely on the project extension and the performance extension. Is the intention that we will move that logic into one of the extensions and check the query before applying logic (specifically, check that the query has `project_ids` before calling `get_projects_query_flags` if we do it in the performance extension)?
1. It also seems like we do some rate limiting based on the extension [here](https://github.com/getsentry/snuba/blob/master/snuba/util.py#L394-L397). That doesn't move into the processor, right?

Reply:
1) we should loop through all the processors
2) This os a good point. It is unfortunate that an extension depends on another. First I would say that the performance extension should work with or without project. Second we can extend the extension schema to include the project_id for performance and have two overlapping schemas. This would allow us not to have the extension processor to depend on something outside its extensions chema
3) the rate limiting logic does not go into the query schema extension. We should actually redesign the rate limit to define what to do if the project is not present in the query, I think this is orthogonal

Reply:
> 2) This os a good point. It is unfortunate that an extension depends on another. First I would say that the performance extension should work with or without project. Second we can extend the extension schema to include the project_id for performance and have two overlapping schemas. This would allow us not to have the extension processor to depend on something outside its extensions chema

One thing that I had considered earlier: I wonder if it'd be more appropriate to consider the performance stuff as query _settings_, separate from extensions. The only thing that kept me from doing this earlier was the increased complexity (it's just adding a third attribute to the `Request` type, so not really a big deal) and wanting to figure out how general of a concept something like "turbo" or "consistent" is/will be before exposing it to all datasets. 

Doing this would make the extensions (presumably) able to use these settings as part of their execution, so that might be worth considering here.

> 3) the rate limiting logic does not go into the query schema extension. We should actually redesign the rate limit to define what to do if the project is not present in the query, I think this is orthogonal 

Totally agree, my thinking was that it might make sense for each dataset to be able to return an iterable of quota objects that are derived from the request. Things like outcomes won't be able to use the project as a basis for rate limiting like we've discussed earlier. I think this will have to happen sooner rather than later. I also think we should generally switch to organization-based rate limits here instead of project rate limits to avoid weird cases like this: https://github.com/getsentry/snuba/blob/84e4aa20afbac870365aafe3ed9c96cb09f589e2/snuba/util.py#L382

Reply:
I would go this way (this PR is already more than 600 lines):
1) after this PR get rid of the performance extension and make that code work with or without project_id
2) after this PR refactor the rate limit. 
How does it sound ?